src/persist/src/storage.rs

```diff
@@ -421,119 +451,164 @@ pub mod tests {
-        assert_eq!(blob0.delete("nope").await, Ok(()));
+        assert_eq!(full0.delete("nope").await, Ok(()));
 
         // Empty blob contains no keys.
-        blob0.delete("k0a").await?;
-        let mut blob_keys = blob0.list_keys().await?;
+        full0.delete("k0a").await?;
+        let mut blob_keys = full0.list_keys().await?;
+        blob_keys.sort();
+        assert_eq!(blob_keys, empty_keys);
+        let mut blob_keys = read0.list_keys().await?;
         blob_keys.sort();
         assert_eq!(blob_keys, empty_keys);
         // Can reset a deleted key to some other value.
-        blob0.set("k0", values[1].clone(), AllowNonAtomic).await?;
-        assert_eq!(blob0.get("k0").await?, Some(values[1].clone()));
+        full0.set("k0", values[1].clone(), AllowNonAtomic).await?;
+        assert_eq!(read0.get("k0").await?, Some(values[1].clone()));
+        assert_eq!(full0.get("k0").await?, Some(values[1].clone()));
 
         // Insert multiple keys back to back and validate that we can list
         // them all out.
         let mut expected_keys = empty_keys;
         for i in 1..=5 {
             let key = format!("k{}", i);
-            blob0.set(&key, values[0].clone(), AllowNonAtomic).await?;
+            full0.set(&key, values[0].clone(), AllowNonAtomic).await?;
             expected_keys.push(key);
         }
 
         // Blob contains the key we just inserted.
-        let mut blob_keys = blob0.list_keys().await?;
+        let mut blob_keys = full0.list_keys().await?;
+        blob_keys.sort();
+        assert_eq!(blob_keys, keys(&expected_keys, &["k0"]));
+        let mut blob_keys = read0.list_keys().await?;
         blob_keys.sort();
         assert_eq!(blob_keys, keys(&expected_keys, &["k0"]));
 
         // Cannot reuse a blob once it is closed.
-        assert_eq!(blob0.close().await, Ok(true));
-        assert!(blob0.get("k0").await.is_err());
-        assert!(blob0
+        assert_eq!(full0.close().await, Ok(true));
+        assert!(full0.get("k0").await.is_err());
+        assert!(full0.list_keys().await.is_err());
+        assert!(full0
             .set("k1", values[0].clone(), RequireAtomic)
             .await
             .is_err());
+        assert!(full0.delete("k0").await.is_err());
 
         // Close must be idempotent and must return false if it did no work.
-        assert_eq!(blob0.close().await, Ok(false));
+        assert_eq!(full0.close().await, Ok(false));
 
-        // But we can reopen it and use it.
-        let blob0 = new_fn(PathAndReentranceId {
+        // Closing the exclusive-writer doesn't affect the availability of any
+        // readers.
+        assert_eq!(read0.get("k0").await?, Some(values[1].clone()));
+
+        // We can reopen the exclusive-writer with a different reentrance_id and
+        // use it.
+        let full0 = new_full_fn(PathAndReentranceId {
             path: "path0",
-            reentrance_id: "reentrance0",
+            reentrance_id: "reentrance1",
```

Comment:
🙈 s3 wasn't deleting the lock file on close and this is why the test didn't catch it. doh!

Reply:
nice!